### ISSUE TITLE
Please do not make nested go.mod modules

### ISSUE THREAD

**Issue by litecomb** [btcsuite/btcd#2271]
Rookies have been active, creating unnecessary nested go modules. These are notoriously [hard to get right](https://github.com/golang/go/issues/59458), cause part of the project depend on entirely different versions of itself, and even though creating nested modules appears ([on](https://github.com/golang/go/issues/64104) [the](https://github.com/golang/go/issues/61237) [surface](https://github.com/golang/go/issues/59741) [only](https://github.com/golang/go/issues/59458)) to cause no problem, the opposite is true. Whats worse, once the problem is spotted, deleting the nested modules is near impossible.

The attached patch in gist attempts to clean the mess caused by other contributors:

[https://gist.github.com/litecomb/ea736ccb1573ee468f40518065f59329](https://gist.github.com/litecomb/ea736ccb1573ee468f40518065f59329)

Note that this assumes that beyond that point the btcec package will share its version with the rest of the repo. I have no idea if that is desired or necessary at all.

Unfortunately since deleting the nested modules is near impossible, a full fork of the repo may be necessary to properly flush external sites such as pkg.go.dev and the infamous google mitm source code proxy which backs go get.

**Comment by guggero** [comment]
I agree that having nested modules that are more than one level deep (e.g. `btcd -> btcutil -> psbt`) isn't ideal.
And there's an [old PR](https://github.com/btcsuite/btcd/pull/1825) that tries to clean that up somewhat, you're welcome to review that.

But not having _any_ submodules isn't an option either. Do you really want to pull in **all** of `btcd` into your project just because you want to use the elliptic curve library (`btcec`) or want to be able to create a BTC address (`btcutil`) in Golang?
Moving those libraries into other repositories doesn't really help a lot either, as it makes development way harder (dependencies across multiple projects). In fact, `btcutil` used to be its own project but was moved here to avoid exactly those development dependency issues.

**Comment by guggero** [comment]
Oh we definitely understand the developer pain that comes with submodules. We've been using them in all our projects and it's definitely not easy.

But, especially with `btcd`, many of the packages within some projects are mostly seen as a **library** first.
So being able to pull in just those library packages through the use of submodules is important.
Especially if the final build size is a concern as is with mobile and WASM builds.

So I'm definitely not saying your concerns aren't valid. I'm just saying that the chances of this project changing its approach are very slim in my opinion.
We have many consumers (especially on the `btcec` library) and almost no reviewers. So a massive code change as you're suggesting will likely not get much review attention (just take a look at my PR linked above that actually _fixes_ issues).

But of course you're welcome to try, I'd just recommend to manage expectations.

**Comment by litecomb** [comment]
Reminded that before tagging v0.25, all go.mod files need attention:

#2386
